© 2022 grdnr13 <grdnr17@gmail.com>



---


## 1-4. Support vector machines with Iris dataset


---



Download the dataset


*   data source: scikit-learn library



In [100]:
from sklearn.datasets import load_iris
 
iris_dataset = load_iris()

## Goal: Train Support vector machines to predict y (species of flower) based on provided features (SepalLength, SepalWidth, PetalLength, PetalWidth)


### Split the dataset into training and testing sets (training:testing=75:25)

In [101]:
#print(iris_dataset)

In [102]:
from sklearn.model_selection import train_test_split

In [103]:
X = iris_dataset['data']
y = iris_dataset['target']
print(X.shape)
print(y.shape)

(150, 4)
(150,)


In [104]:
#split
X_train, X_test = train_test_split(X, test_size=0.25, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.25, random_state=0)
print(X_train.shape, X_test.shape)
print(y_train.shape, y_test.shape)


(112, 4) (38, 4)
(112,) (38,)


### Use "StandardScaler" function (provided in HA1_part1) to map all points to have mean=0 and std=1

In [105]:
from sklearn.preprocessing import StandardScaler
def scaler_samples(train_X,test_X):
  scaler = StandardScaler()
  train_X = scaler.fit_transform(train_X)
  test_X = scaler.transform(test_X)

  return train_X, test_X

In [106]:
scaler_samples(X_train, X_test)

(array([[ 1.54399532e-02, -1.19254753e-01,  2.25126850e-01,
          3.55797625e-01],
        [-9.98450310e-02, -1.04039491e+00,  1.13559562e-01,
         -2.98410911e-02],
        [ 1.05300481e+00, -1.19254753e-01,  9.50314227e-01,
          1.12707506e+00],
        [-1.36797986e+00,  3.41315328e-01, -1.39259884e+00,
         -1.31530348e+00],
        [ 1.16828980e+00,  1.11030287e-01,  7.27179649e-01,
          1.38416753e+00],
        [-1.02212490e+00,  1.03217045e+00, -1.22524790e+00,
         -8.01118523e-01],
        [-5.60984968e-01,  1.49274053e+00, -1.28103155e+00,
         -1.31530348e+00],
        [-1.02212490e+00, -2.42210516e+00, -1.65358660e-01,
         -2.86933568e-01],
        [ 7.07149859e-01, -1.19254753e-01,  9.50314227e-01,
          7.41436341e-01],
        [ 9.37719827e-01,  5.71600368e-01,  1.06188152e+00,
          1.64126001e+00],
        [ 1.30724937e-01, -1.96153508e+00,  6.71396005e-01,
          3.55797625e-01],
        [ 9.37719827e-01, -1.27067995e+00, 

### First, train and test SVM using linear kernel
- Report the accuracy
- After that, by using 5-fold cross validation, again report accuracies from each fold and the mean accuracy after taking the mean on five accuracies.

In [107]:
from sklearn.svm import SVC
model_linear=SVC(kernel="linear")

#train
model_linear.fit(X_train, y_train)


SVC(kernel='linear')

In [108]:
from sklearn.metrics import accuracy_score
#accuracy
predictions_linear = model_linear.predict(X_test)
print(accuracy_score(y_test, predictions_linear))


0.9736842105263158


In [109]:
# print(model_linear.support_vectors_)

In [110]:
# by using 5-fold cross validation, again report accuracies from each fold 
# and the mean accuracy after taking the mean on five accuracies.
from sklearn.model_selection import cross_val_score

scores_linear = cross_val_score(model_linear, X_test, y_test, cv=5)

print('5-fold cross validation score\n{}'.format(scores_linear))
print('mean of cross validation score\n{:.3f}'.format(scores_linear.mean()))


5-fold cross validation score
[1.         1.         0.875      0.85714286 0.85714286]
mean of cross validation score
0.918


### Second, train and test SVM using RBF kernel
use C=1, gamma=0.01
- Report the accuracy
- After that, by using 5-fold cross validation, again report accuracies from each fold and the mean accuracy after taking the mean on five accuracies.

In [111]:
model_rbf=SVC(kernel="rbf", gamma=0.01, C=1)

#train
model_rbf.fit(X_train, y_train)


SVC(C=1, gamma=0.01)

In [112]:
#accuracy
predictions_rbf = model_rbf.predict(X_test)
print(accuracy_score(y_test, predictions_rbf))

0.9210526315789473


In [113]:
# by using 5-fold cross validation, again report accuracies from each fold
# and the mean accuracy after taking the mean on five accuracies.
scores_rbf = cross_val_score(model_rbf, X_test, y_test, cv=5)

print('5-fold cross validation score\n{}'.format(scores_rbf))
print('mean of cross validation score\n{:.3f}'.format(scores_rbf.mean()))

5-fold cross validation score
[0.75       0.75       0.75       0.71428571 0.85714286]
mean of cross validation score
0.764
5-fold cross validation score
[0.75       0.75       0.75       0.71428571 0.85714286]
mean of cross validation score
0.764


### Hyperparemter tuning by using Grid Search (i.e., from sklearn.model_selection import GridSearchCV)
- Parameter search space must include three kernels (['rbf', 'poly', 'sigmoid'], C ([0.1,1, 10, 100]) and gamma ([1,0.1,0.01,0.001]))
- Report best parameters and accuracy based on them

In [114]:
from sklearn.model_selection import GridSearchCV

# defining parameter range
param_grid = {'C': [0.1, 1, 10, 100, 1000],
'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
'kernel': ['rbf']}
grid = GridSearchCV(SVC(), param_grid, refit = True)

grid.fit(X_train, y_train)


GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.1, 1, 10, 100, 1000],
                         'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
                         'kernel': ['rbf']})

In [115]:
sorted(grid.cv_results_.keys())


['mean_fit_time',
 'mean_score_time',
 'mean_test_score',
 'param_C',
 'param_gamma',
 'param_kernel',
 'params',
 'rank_test_score',
 'split0_test_score',
 'split1_test_score',
 'split2_test_score',
 'split3_test_score',
 'split4_test_score',
 'std_fit_time',
 'std_score_time',
 'std_test_score']

In [116]:
print('Parameter:', grid.best_params_)
print('Accuracy: {:.4f}'.format(grid.best_score_))

Parameter: {'C': 10, 'gamma': 0.1, 'kernel': 'rbf'}
Accuracy: 0.9731


### Extend your discussion on why chosen kernel works more effectively than others and whether hyperparmeter tuning helps perfomrance improvement.


Hyperparameter tuning using grid search work showed accuracy of 97.31%, while SVM using linear kernel showed 91.8% accuracy, and SVM using RBF kernel showed 76.4% accuracy.


Therefore, we can see that hyperparameter tuning helps performance improvement.